File: src/noitu_bot.py
import random
import db

# Load danh sách các từ có 2 từ vào một list
with open('src/assets/tudien.txt', 'r') as f:
    list_words = [word.strip().lower() for word in f.readlines()]


def last_word(word):
    return word.split()[-1]

# trích xuất từ đầu tiên của một từ


def first_word(word):
    return word.split()[0]


def get_word_starting_with(start):
    matching_words = [word for word in list_words if word.split()[0] == start]
    if matching_words:
        word = random.choice(matching_words)
        return word
    else:
        return False


def unique_word(start):
    matching_words = [word for word in list_words if word.split()[0] == start]
    return True if len(matching_words) == 1 else False


def check_channel(player_word, id_channel, id_user):
    return 'test'
    global sai, current_word, data
    id = str(id)
    if id in data["word"]:
        print('ok')
    else:
        print('ok 2')
        return start()

    if not current_word:
        current_word = random.choice(list_words)

    if last_word(current_word) == first_word(player_word) and sai != 1:
        if player_word in history:
            return 'Đã trả lời từ, vui lòng tìm từ khác'
        if player_word in list_words:
            # Tìm một từ mới từ danh sách các từ có 2 từ để đưa ra
            next_word = get_word_starting_with(last_word(player_word))
            current_word = next_word
            if not next_word:
                return win()
            response = 'Từ tiếp theo: **' + next_word + '**'
            return response
        else:
            print('Không tồn tại từ, vui lòng tìm từ khác')
            sai -= 1
            response = 'Không tồn tại từ, vui lòng tìm từ khác, **còn ' + \
                str(sai) + ' lần thử** \nTừ hiện tại: **' + current_word + '**'
            return response
    # else:
    #     return loss()



def check_user(player_word, id_user):
    id_user = str(id_user)
    user_data = db.read('users').get(id_user, {})
    current_word = user_data.get('word')
    history = user_data.get('history', [])
    streak = user_data.get('streak', 0)
    sai = user_data.get('sai', 0)

    # nếu không tồn tại user thì tạo user mới
    if not current_word:
        current_word = random.choice(list_words)
        db.store('users', {id_user: {'word': current_word,
                 'history': [current_word], 'streak': 0, 'sai': 0}})
        return f'Từ hiện tại: **{current_word}**'

    # nếu từ cuối không giống từ đầu
    if last_word(current_word) != first_word(player_word):
        sai += 1
        if sai == 3:
            current_word = random.choice(list_words)
            db.store('users', {id_user: {'word': current_word,
                     'history': [], 'streak': 0, 'sai': 0}})
            return f'> Thua cuộc, từ đầu bạn đưa ra phải trùng với từ cuối của bot hoặc từ phải có nghĩa! Chuỗi đúng: **{streak}** \nTừ mới: **{current_word}**'
        else:
            db.store('users', {id_user: {'word': current_word,
                     'history': history, 'streak': streak, 'sai': sai}})
            return f'Đã trả lời sai, vui lòng tìm từ khác. Bạn đã trả lời sai {sai} lần.\nTừ hiện tại: **{current_word}**'

    # nếu từ không có hoặc đã trả lời
    if player_word in history or player_word not in list_words:
        sai += 1
        if sai == 3:
            current_word = random.choice(list_words)
            db.store('users', {id_user: {'word': current_word,
                     'history': [], 'streak': 0, 'sai': 0}})
            return f'> Thua cuộc, từ của bạn không tồn tại hoặc đã trả lời! Chuỗi đúng: **{streak}** \nTừ mới: **{current_word}**'
        else:
            db.store('users', {id_user: {'word': current_word,
                     'history': history, 'streak': streak, 'sai': sai}})
            return f'Đã trả lời từ hoặc từ không hợp lệ, vui lòng tìm từ khác. Bạn đã trả lời sai {sai} lần.\nTừ hiện tại: **{current_word}**'

    next_word = get_word_starting_with(last_word(player_word))
    current_word = next_word

    # nếu từ tiếp theo không có trong từ điển
    if not next_word and not next_word in history:
        current_word = random.choice(list_words)
        db.store('users', {id_user: {'word': current_word,
                 'history': [], 'streak': 0, 'sai': 0}})
        return f'**BẠN ĐÃ THẮNG!** \nChuỗi đúng: **{streak+1}**\n Từ mới: **{current_word}**'

    response = f'Từ tiếp theo: **{next_word}**'

    # nếu đây là từ duy nhất có thể trả lời
    if unique_word(first_word(next_word)):
        current_word = random.choice(list_words)
        db.store('users', {id_user: {'word': current_word,
                 'history': [], 'streak': 0, 'sai': 0}})
        return f'{response}\n> Thua cuộc, đây là từ cuối trong từ điển của bot, Chuỗi đúng: **{streak}**\nTừ mới: **{current_word}**'

    # nếu không thì ra từ tiếp theo và thêm vào lịch sử
    history.extend([player_word, current_word])
    db.store(
        'users', {id_user: {'word': current_word, 'history': history, 'streak': streak+1}})
    return response

File: src/noitu_bot.js
const { listWords, wordPairs, normalizeVietnamese } = require('./noitu');
const db = require('./db');
const { setupLogger } = require('./log');

const logger = setupLogger('noitu_bot');

function lastWord(word) {
  return word.split(' ')[word.split(' ').length - 1];
}

function firstWord(word) {
  return word.split(' ')[0];
}

function getWordStartingWith(start, history = []) {
  const possibleWords = wordPairs[start] || [];
  
  if (possibleWords.length === 0) {
    return false;
  }
  
  // Lọc các từ chưa được sử dụng trong lịch sử
  const availableWords = possibleWords.filter(secondWord => {
    const fullWord = `${start} ${secondWord}`;
    return !history.includes(fullWord);
  });
  
  // Nếu không còn từ nào có thể dùng, trả về false
  if (availableWords.length === 0) {
    return false;
  }
  
  // Tránh chọn từ giống nhau (ví dụ: "phới phới")
  const nonRepeatingWords = availableWords.filter(secondWord => secondWord !== start);
  
  // Ưu tiên từ không lặp, nếu không có thì dùng từ có sẵn
  const wordsToChoose = nonRepeatingWords.length > 0 ? nonRepeatingWords : availableWords;
  
  const secondWord = wordsToChoose[Math.floor(Math.random() * wordsToChoose.length)];
  return `${start} ${secondWord}`;
}

function uniqueWord(start) {
  const possibleWords = wordPairs[start] || [];
  return possibleWords.length === 0;
}

function newWord() {
  let word;
  do {
    word = listWords[Math.floor(Math.random() * listWords.length)];
  } while (uniqueWord(lastWord(word)));
  return word;
}

function checkChannel(playerWord, idChannel, idUser) {
  const startTime = Date.now();
  idChannel = idChannel.toString();

  const normalizedPlayer = normalizeVietnamese(playerWord);

  if (normalizedPlayer.split(' ').length !== 2) {
    return 'Từ bắt buộc phải gồm 2 từ';
  }

  const channels = db.read('channels') || {};
  let channelData = channels[idChannel] || {};
  let currentWord = channelData.word;
  let history = channelData.history || [];
  let streak = channelData.streak || 0;
  let sai = channelData.sai || 0;

  if (!currentWord) {
    currentWord = newWord();
    channelData = { word: currentWord, history: [currentWord], streak: 0, sai: 0 };
    db.store('channels', { [idChannel]: channelData });
    logger.info(`Channel [${idChannel}] NEW '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
    return `Từ hiện tại: **${currentWord}**`;
  }

  if (lastWord(currentWord) !== firstWord(normalizedPlayer)) {
    sai += 1;
    if (sai === 3) {
      currentWord = newWord();
      channelData = { word: currentWord, history: [], streak: 0, sai: 0 };
      db.store('channels', { [idChannel]: channelData });
      logger.info(`Channel [${idChannel}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> Thua cuộc, từ đầu không khớp với từ cuối! Chuỗi đúng: **${streak}** \nTừ mới: **${currentWord}**`;
    } else {
      channelData.sai = sai;
      db.store('channels', { [idChannel]: channelData });
      logger.info(`Channel [${idChannel}] ERROR '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> **Từ đầu của bạn phải là "${lastWord(currentWord)}"!** Vui lòng tìm từ khác. Bạn đã trả lời sai **${sai}** lần.\nTừ hiện tại: **${currentWord}**`;
    }
  }

  // Kiểm tra từ đã được trả lời chưa
  if (history.includes(normalizedPlayer)) {
    sai += 1;
    if (sai === 3) {
      currentWord = newWord();
      channelData = { word: currentWord, history: [], streak: 0, sai: 0 };
      db.store('channels', { [idChannel]: channelData });
      logger.info(`Channel [${idChannel}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> Thua cuộc, từ đã được trả lời trước đó! Chuỗi đúng: **${streak}** \nTừ mới: **${currentWord}**`;
    } else {
      channelData.sai = sai;
      db.store('channels', { [idChannel]: channelData });
      logger.info(`Channel [${idChannel}] ERROR '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> **Từ này đã được trả lời trước đó!** Vui lòng tìm từ khác. Bạn đã trả lời sai **${sai}** lần.\nTừ hiện tại: **${currentWord}**`;
    }
  }

  // Kiểm tra từ có trong từ điển không
  if (!listWords.includes(normalizedPlayer)) {
    sai += 1;
    if (sai === 3) {
      currentWord = newWord();
      channelData = { word: currentWord, history: [], streak: 0, sai: 0 };
      db.store('channels', { [idChannel]: channelData });
      logger.info(`Channel [${idChannel}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> Thua cuộc, từ không có trong bộ từ điển! Chuỗi đúng: **${streak}** \nTừ mới: **${currentWord}**`;
    } else {
      channelData.sai = sai;
      db.store('channels', { [idChannel]: channelData });
      logger.info(`Channel [${idChannel}] ERROR '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> **Từ không có trong bộ từ điển!** Vui lòng tìm từ khác. Bạn đã trả lời sai **${sai}** lần.\nTừ hiện tại: **${currentWord}**`;
    }
  }

  const nextWord = getWordStartingWith(lastWord(normalizedPlayer), history);
  currentWord = nextWord;

  if (!nextWord) {
    currentWord = newWord();
    channelData = { word: currentWord, history: [], streak: 0, sai: 0 };
    db.store('channels', { [idChannel]: channelData });
    logger.info(`Channel [${idChannel}] WIN '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
    return `**BẠN ĐÃ THẮNG!** \nChuỗi đúng: **${streak + 1}**\n Từ mới: **${currentWord}**`;
  }

  const response = `Từ tiếp theo: **${nextWord}**`;

  if (uniqueWord(lastWord(nextWord))) {
    currentWord = newWord();
    channelData = { word: currentWord, history: [], streak: 0, sai: 0 };
    db.store('channels', { [idChannel]: channelData });
    logger.info(`Channel [${idChannel}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
    return `${response}\n> Thua cuộc, đây là từ cuối trong từ điển của bot, Chuỗi đúng: **${streak}**\nTừ mới: **${currentWord}**`;
  }

  history.push(normalizedPlayer, currentWord);
  channelData.word = currentWord;
  channelData.history = history;
  channelData.streak = streak + 1;
  db.store('channels', { [idChannel]: channelData });
  logger.info(`Channel [${idChannel}] NEXT '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
  return response;
}

function checkUser(playerWord, idUser) {
  const startTime = Date.now();
  idUser = idUser.toString();

  if (playerWord.split(' ').length !== 2) {
    return 'Từ bắt buộc phải gồm 2 từ';
  }

  const users = db.read('users') || {};
  let userData = users[idUser] || {};
  let currentWord = userData.word;
  let history = userData.history || [];
  let streak = userData.streak || 0;
  let sai = userData.sai || 0;

  if (!currentWord) {
    currentWord = newWord();
    userData = { word: currentWord, history: [currentWord], streak: 0, sai: 0 };
    db.store('users', { [idUser]: userData });
    logger.info(`DM: [${idUser}] NEW '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
    return `Từ hiện tại: **${currentWord}**`;
  }

  if (lastWord(currentWord) !== firstWord(playerWord)) {
    sai += 1;
    if (sai === 3) {
      currentWord = newWord();
      userData = { word: currentWord, history: [], streak: 0, sai: 0 };
      db.store('users', { [idUser]: userData });
      logger.info(`DM: [${idUser}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> Thua cuộc, từ đầu không khớp với từ cuối! Chuỗi đúng: **${streak}** \nTừ mới: **${currentWord}**`;
    } else {
      userData.sai = sai;
      db.store('users', { [idUser]: userData });
      logger.info(`DM: [${idUser}] ERROR '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> **Từ đầu của bạn phải là "${lastWord(currentWord)}"!** Vui lòng tìm từ khác. Bạn đã trả lời sai **${sai}** lần.\nTừ hiện tại: **${currentWord}**`;
    }
  }

  // Kiểm tra từ đã được trả lời chưa
  if (history.includes(playerWord)) {
    sai += 1;
    if (sai === 3) {
      currentWord = newWord();
      userData = { word: currentWord, history: [], streak: 0, sai: 0 };
      db.store('users', { [idUser]: userData });
      logger.info(`DM: [${idUser}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> Thua cuộc, từ đã được trả lời trước đó! Chuỗi đúng: **${streak}** \nTừ mới: **${currentWord}**`;
    } else {
      userData.sai = sai;
      db.store('users', { [idUser]: userData });
      logger.info(`DM: [${idUser}] ERROR '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> **Từ này đã được trả lời trước đó!** Vui lòng tìm từ khác. Bạn đã trả lời sai **${sai}** lần.\nTừ hiện tại: **${currentWord}**`;
    }
  }

  // Kiểm tra từ có trong từ điển không
  if (!listWords.includes(playerWord)) {
    sai += 1;
    if (sai === 3) {
      currentWord = newWord();
      userData = { word: currentWord, history: [], streak: 0, sai: 0 };
      db.store('users', { [idUser]: userData });
      logger.info(`DM: [${idUser}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> Thua cuộc, từ không có trong bộ từ điển! Chuỗi đúng: **${streak}** \nTừ mới: **${currentWord}**`;
    } else {
      userData.sai = sai;
      db.store('users', { [idUser]: userData });
      logger.info(`DM: [${idUser}] ERROR '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
      return `> **Từ không có trong bộ từ điển!** Vui lòng tìm từ khác. Bạn đã trả lời sai **${sai}** lần.\nTừ hiện tại: **${currentWord}**`;
    }
  }

  const nextWord = getWordStartingWith(lastWord(playerWord), history);
  currentWord = nextWord;

  if (!nextWord) {
    currentWord = newWord();
    userData = { word: currentWord, history: [], streak: 0, sai: 0 };
    db.store('users', { [idUser]: userData });
    logger.info(`DM: [${idUser}] WIN '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
    return `**BẠN ĐÃ THẮNG!** \nChuỗi đúng: **${streak + 1}**\n Từ mới: **${currentWord}**`;
  }

  const response = `Từ tiếp theo: **${nextWord}**`;

  if (uniqueWord(lastWord(nextWord))) {
    currentWord = newWord();
    userData = { word: currentWord, history: [], streak: 0, sai: 0 };
    db.store('users', { [idUser]: userData });
    logger.info(`DM: [${idUser}] LOSS '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
    return `${response}\n> Thua cuộc, đây là từ cuối trong từ điển của bot, Chuỗi đúng: **${streak}**\nTừ mới: **${currentWord}**`;
  }

  history.push(playerWord, currentWord);
  userData.word = currentWord;
  userData.history = history;
  userData.streak = streak + 1;
  db.store('users', { [idUser]: userData });
  logger.info(`DM: [${idUser}] NEXT '${playerWord}' -> '${currentWord}' [${(Date.now() - startTime) / 1000}s]`);
  return response;
}

async function tratu() {
  // Placeholder
  return "Chức năng tra từ";
}

function resetUserGame(idUser) {
  idUser = idUser.toString();
  const currentWord = newWord();
  const userData = { word: currentWord, history: [currentWord], streak: 0, sai: 0 };
  db.store('users', { [idUser]: userData });
  logger.info(`Reset user game for [${idUser}], new word: ${currentWord}`);
  return currentWord;
}

function resetChannelGame(idChannel) {
  idChannel = idChannel.toString();
  const currentWord = newWord();
  const channelData = { word: currentWord, history: [currentWord], streak: 0, sai: 0 };
  db.store('channels', { [idChannel]: channelData });
  logger.info(`Reset channel game for [${idChannel}], new word: ${currentWord}`);
  return currentWord;
}

module.exports = { checkChannel, checkUser, tratu, resetUserGame, resetChannelGame };